### ISSUE TITLE
Uplift clippy::mem_discriminant_non_enum into rustc

### ISSUE THREAD

**Issue by Flying-Toast** [rust-lang/rust#83899]
The [clippy::mem_discriminant_non_enum](https://rust-lang.github.io/rust-clippy/master/index.html#mem_discriminant_non_enum) lint seems like it should be directly in rustc. The lint prevents doing something like:
```Rust
core::mem::discriminant::<i32>(&4)
```
which causes unspecified behavior.

**Comment by Flying-Toast** [comment]
I'd like to try implementing this myself, if would be likely to be accepted - do people think this would this be useful?

**Comment by jyn514** [comment]
I think this is a great idea, it's clearly wrong code with no false positives.

**Comment by Flying-Toast** [comment]
Should I make this a hard error instead of a lint?

**Comment by jyn514** [comment]
No, I think this should be a lint. Hard errors are a breaking change and are almost never accepted.
